tensorflow-frameworks/tensorflow-data/src/main/java/org/tensorflow/data/Dataset.java

```diff
@@ -0,0 +1,180 @@
+package org.tensorflow.data;
+
+import org.tensorflow.*;
+import org.tensorflow.data.impl.*;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.data.AnonymousIterator;
+import org.tensorflow.op.data.MakeIterator;
+import org.tensorflow.tools.Shape;
+import org.tensorflow.utils.Tuple2;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a potentially large list of independent elements (samples), and
+ * allows iteration and transformations to be performed across these elements.
+ */
+public abstract class Dataset implements Iterable<List<Output<?>>> {
+  protected Ops tf;
+  private List<DataType<?>> outputTypes;
+  private List<Shape> outputShapes;
+
+  public Dataset(Ops tf, List<DataType<?>> outputTypes, List<Shape> outputShapes) {
+    if (Objects.isNull(tf)) {
+      throw new IllegalArgumentException("Ops accessor cannot be null.");
+    } else if (outputTypes.size() != outputShapes.size()) {
+      throw new IllegalArgumentException("`outputTypes` and `outputShapes` must have the same size.");
+    }
+
+    this.tf = tf;
+    this.outputTypes = outputTypes;
+    this.outputShapes = outputShapes;
+  }
+
+  /**
+   * Groups elements of this dataset into batches.
+   *
+   * @param batchSize     The number of desired elements per batch
+   * @param dropLastBatch Whether to leave out the final batch if it has fewer
+   *                      than `batchSize` elements.
+   * @return A batched Dataset
+   */
+  public final Dataset batch(long batchSize, boolean dropLastBatch) {
+    List<Shape> batchOutputShapes = getOutputShapes().stream()
+        .map(s -> Shape.of(Utils.array(batchSize, s.asArray())))
+        .collect(Collectors.toList());

```

Comment:
I don't know how `tf.data.batchDataset` works but I find it strange that we need to add the `batchSize` as a new dimension for all individual shapes. Just saying in case this is a mistake

Reply:
This isn't a mistake, it does seem to be needed here, since it's the `outputShapes` that need to be specified.